### PR TITLE
Tests v3

### LINKED ISSUE #738: Call to unit tests generates unnecessary warnings
Some of our unit test checks are generating warnings. For example 
`pyccl/tests/test_halomodel.py::test_halomodel_power_consistent
  /Users/Chisa002/Documents/LSST/CCL/pyccl/pyutils.py:403: CCLWarning: The function twohalo_matter_power is deprecated. Use halomod_power_spectrum instead.
    warnings.warn(s, CCLWarning)`
This can be avoided by updating the functions used in those tests to the new ones.

### PR DESCRIPTION
**Pull request by nikfilippas**
This PR:
- implements v3 in testing for `halos/` (I didn't update the tests there to avoid clogging review even more),
- removes all warnings from testing (Closes #738  ),
- replaces `numpy.testing` and the custom `assert_` functions in `pyutils` with the industry-standard `pytest`.

Reviewing this one should be easy as there are no changed tests whatsoever.

### CODE REVIEW

File: pyccl/halos/massdef.py [hunk 1]
```diff
@@ -6,6 +6,11 @@
 from functools import cached_property
 
 
+__all__ = ("mass2radius_lagrangian", "convert_concentration", "MassDef",
+           "MassDef200m", "MassDef200c", "MassDef500c", "MassDefVir",
+           "MassDefFof",)
+
+
```
Comment (damonge) [inline pyccl/halos/massdef.py:13]: This now appears twice

File: pyccl/base/schema.py [hunk 2]
```diff
@@ -136,6 +136,27 @@ def Funlock(cls, cl, name, mutate: bool):
...
+                if this_name is init_name:  # skip this if called via `super()`
+                    self._object_lock.lock()
+                return out
+            return wrapper
+
+        func = vars(cl).get(name)
+        if func is not None:
+            newfunc = lock_on_exit(func)
+            setattr(cl, name, newfunc)
+
```
Comment (damonge) [inline pyccl/base/schema.py:159]: At this point I've lost track of how many times we've lost and brought back these functions. Please explain why we need to bring this back (or in)

**Review by damonge — COMMENTED** [review]
Ok, quick review

File: pyccl/base/schema.py (hunk 2 above)
Comment (nikfilippas) [reply pyccl/base/schema.py:159]: bad merge, fixed, this shouldn't be there

File: pyccl/halos/massdef.py (hunk 1 above)
Comment (nikfilippas) [reply pyccl/halos/massdef.py]: fixed

**Review by damonge — APPROVED** [review]
LGTM!

**Pull request merged**